### ISSUE TITLE
Proposal: re.prefixmatch method (alias for re.match)

### ISSUE THREAD

**Issue by gpshead** [python/cpython#86519]
BPO | [42353](https://bugs.python.org/issue42353)
--- | :---
Nosy | @gpshead, @ezio-melotti, @serhiy-storchaka, @msuozzo
PRs | <li>python/cpython#31137</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/gpshead'
closed_at = None
created_at = <Date 2020-11-13.19:27:03.105>
labels = ['expert-regex', 'type-feature', 'library', '3.11']
title = 'Proposal: re.prefixmatch method (alias for re.match)'
updated_at = <Date 2022-02-05.08:57:42.604>
user = 'https://github.com/gpshead'
```

bugs.python.org fields:
```python
activity = <Date 2022-02-05.08:57:42.604>
actor = 'serhiy.storchaka'
assignee = 'gregory.p.smith'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)', 'Regular Expressions']
creation = <Date 2020-11-13.19:27:03.105>
creator = 'gregory.p.smith'
dependencies = []
files = []
hgrepos = []
issue_num = 42353
keywords = ['patch']
message_count = 6.0
messages = ['380928', '380975', '380984', '380996', '412554', '412564']
nosy_count = 5.0
nosy_names = ['gregory.p.smith', 'ezio.melotti', 'mrabarnett', 'serhiy.storchaka', 'matthew.suozzo']
pr_nums = ['31137']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue42353'
versions = ['Python 3.11']
```

</p></details>

### Linked PRs
* gh-148096

**Comment by gpshead** [comment]
From an explicit is better than implicit standpoint I still think this is a good idea. **Even if never actually deprecating and removing `re.match`**

Just promoting use of the explicit spelling of the API call in documentation and examples should be benificial to the ecosystem as a whole. Code analyzers (pylint, etc) could then suggest `prefixmatch` as more explicit when it finds an `re.match` call.

_(I edited the opening message to make it more clear, yay markdown)_

**Comment by ezio-melotti** [comment]
I'm also not convinced :thinking:

> A well known anti-pattern in Python is use of `re.match` when you meant to use `re.search`.

If people are not aware that `re.prefixmatch` exists and/or of the behavior of `re.match`, then they will keep using `re.match`, possibly incorrectly.  Many are not aware of `re.fullmatch` either.

> The goal would be to allow linters to ultimately flag `re.match` as the anti-pattern when in 3.12+ mode. Asking people to use `re.prefixmatch` or `re.search` instead.

This will only reach people that use such linters, similarly to how a warning in the docs only reaches people that do read the docs (and notice the warning).  Both could be applied to mitigate the problem, but neither are definitive solutions, and adding a new function has a cost.  So, unless I'm missing something, the addition will only serve to provide a more explicit alternative for the subset of users that use a linter or that are aware of the function and decide to use it.

However, an alternative already exist: they could just use `re.search` in most cases[^1], possibly with a regex like `^...` (if they want the `re.match` behavior) or `^...$` (if they want the `re.fullmatch` behavior).  Technically linters could already suggest that (even though it might be a tougher sell than an explicit function).

> This would help avoid bugs where people mean re.search but write re.match and would ease the cognitive load [...]

It might also increase it though, since now people will have to decide whether to use `search`, `match`, `fullmatch`, `prefixmatch` and wonder about their differences, and the fact that two have different names but the same behavior is potentially confusing (especially if they end up being used together while updating old code).  Some might also wonder why there is no `suffixmatch`, and possibly propose it.

[^1]: there are some difference when `re.MULTILINE` is used: https://docs.python.org/3/library/re.html#search-vs-match

**Comment by gpshead** [comment]
> the addition will only serve to provide a more explicit alternative for the subset of users that use a linter or that are aware of the function and decide to use it.

Yes. Exactly! That is the entire point. To allow those who realize that being explicit reduces the cognitive correctness burden for all developers working in their codebase, huge or small, to do so.

> an alternative already exist: they could just use re.search in most cases, possibly with a regex like ^... (if they want the re.match behavior) or ^...$ (if they want the re.fullmatch behavior).

`^` is not the logical re equivalent to `.match` behavior. There's a different backslash letter thing that is "start of string" which virtually nobody knows off the top of their head so would be more bewildering. Again, it's about the ease of being explicit. It is obvious what behavior `.prefixmatch` implies. It is not at all obvious that `.match` does that given that one talks about regular expression "matches" in English all the time as a matter of terminology and **never** means "oh, only at the start of the string" in that discussion context. Python has a mismatch with real terminology by differing from other major regex sporting languages/libraries used across the industry here. This offers a way out.

Without disruption of removing a very widely used API.

I'm just seeking to enable being explicit in code instead of requiring a code comment saying `# prefixmatch` upon every occurrence of `.match(...)` in code as a preventative measure. Practicality beats purity.

I don't see an added cognitive burden here from a new explicit method name. Nobody needs lookup `prefixmatch`, it's behavior is there in the name. If it's existence causes them question reality and lookup `match`, they'll learn something from the suggestion not to use that API name and perhaps eventually modernize their code.

I would not have filed this feature request if this were it a repeated source of bugs or confusion from those new to Python at work where people regularly use more than one programming language.

Our old terminology has been outvoted by JavaScript, Golang, Rust, and libpcre. We should adapt by allowing developers to be readably explicit about their intent.

**Comment by serhiy-storchaka** [comment]
I have yet to see an example of mistakenly using `re.match()` instead of `re.search()`. `re.search()` can be mistakenly used instead of `re.match()` or `re.fullmatch()`, I seen more than one examples of it, but not vice versa. I can also expect that `re.match()` is used instead of `re.fullmatch()` in a number of cases, mostly in a code that predates `re.fullmatch()`.

I think there is a problem with an initial premise about  "a well known anti-pattern".

We have a habit of removing long deprecated features in Python. For sure, if something is deprecated, it will be removed in 8 or 12 years, even if it is a simple alias. Forcing people to rewrite tons of working code for stylish reason does not look as a worthy goal to me.

**Comment by gpshead** [comment]
I think our disconnect is merely the difference between samples: "you're is not seeing it" and "my colleagues who routinely do code reviews for hundreds of people sometimes new to Python do see it" are both true statements.

**Comment by gpshead** [comment]
Updating `regex` is tracked in https://github.com/mrabarnett/mrab-regex/issues/601
